pkg/chunk/chunk_store.go

```diff
@@ -404,3 +446,23 @@ func (c *store) convertChunkIDsToChunks(ctx context.Context, chunkIDs []string)
 
 	return chunkSet, nil
 }
+
+// We return the stringified entry and the map from that string->entry
+// We cannot safely go back from string to entry hence we need the map.
+func keysFromEntries(entries []IndexEntry) ([]string, map[string]IndexEntry) {
+	keys := make([]string, 0, len(entries))
+	keyMap := make(map[string]IndexEntry, len(entries))
+	for _, entry := range entries {
+		key := strings.Join([]string{
+			entry.TableName,
+			entry.HashValue,
+			string(entry.RangeValue),
+			string(entry.Value),
+		}, string('\xff'))
+
+		keys = append(keys, key)
+		keyMap[key] = entry
+	}
+
+	return keys, keyMap
+}
```

Comment:
I'd be tempted to inline this in `dedupeEntriesFromCache` as its only used there.